tools/shared.py

```diff
@@ -2555,19 +2555,38 @@ def add_to_path(dirname):
       logger.debug('closure compiler: ' + ' '.join(cmd))
 
       proc = run_process(cmd, stderr=PIPE, check=False, env=env)
-      if proc.returncode != 0:
-        sys.stderr.write(proc.stderr)
-        hint = ''
-        if not pretty:
-          hint = ' the error message may be clearer with -g1'
-        exit_with_error('closure compiler failed (rc: %d.%s)', proc.returncode, hint)
 
       # XXX Closure bug: if Closure is invoked with --create_source_map, Closure should create a
       # outfile.map source map file (https://github.com/google/closure-compiler/wiki/Source-Maps)
       # But it looks like it creates such files on Linux(?) even without setting that command line
       # flag (and currently we don't), so delete the produced source map file to not leak files in
       # temp directory.
       try_delete(outfile + '.map')
+
+      if os.getenv('EMCC_DEBUG') and (proc.returncode != 0 or len(proc.stderr.strip()) > 0):
+        logger.debug(open(filename, 'r').read())
+
+      if proc.returncode != 0:
+        logger.error('Closure compiler run failed:\n')
+        logger.error(proc.stderr)
+        exit_with_error('closure compiler failed (rc: %d.%s)', proc.returncode, '' if pretty else ' the error message may be clearer with -g1 and EMCC_DEBUG=1 set')
+
+      if len(proc.stderr.strip()) > 0:
+        if Settings.CLOSURE_WARNINGS_ARE_ERRORS:
+          logger.error('Closure compiler completed with warnings and -s CLOSURE_WARNINGS_ARE_ERRORS=1 enabled, aborting!\n')
+          logger.error(proc.stderr)
+        else:
+          logger.warn('Closure compiler completed with warnings:\n')
+          logger.warn(proc.stderr)
```

Comment:
It seems like all code paths here print proc.stderr so maybe just lift it out ?

Reply:
I don't think this kind of polishing optimizations are very useful, that requires capturing either logger.error or logger.warn to a variable and then invoking that, and it makes the code look clumsy and nonidiomatic in my opinion.

Reply:
OK, I'm just a little sad to see so much logic here for something I hoped would be simple.   We don't do anything complex like this for other subtools like the compiler and linker.

lgtm if you think the output readability if worth if in this case.